keps/sig-cli/2590-kubectl-subresource/README.md

```diff
@@ -0,0 +1,378 @@
+
+If `--subresource` flag is used for a resource that doesn't support the subresource, a `NotFound` error will be returned.
+
+### Examples
+
+#### get
+
+```shell=
+# for built-in types
+$ kubectl get deployment nginx-deployment --subresource=status
+NAME               READY   UP-TO-DATE   AVAILABLE   AGE
+nginx-deployment   3/3     3            3           43s
+
+$ kubectl get deployment nginx-deployment --subresource=scale
+NAME               DESIREDREPLICAS   AVAILABLEREPLICAS
+nginx-deployment   3                 3
+
+# for CRDS
+$ kubectl get crontab cron--subresource=status
+NAME   SPEC          REPLICAS   AGE
+cron   * * * * */5   3          4m52s
+
+$ kubectl get vmset vmset-1 --subresource=scale
+NAME   DESIREDREPLICAS   AVAILABLEREPLICAS
+cron   3                 0
+```
+
+Invalid subresources:
+
+```shell=
+$ kubectl get pod nginx-deployment-66b6c48dd5-dv6gl --subresource=logs
+error: --subresource must be one of [status scale], not "logs"
+
+$ kubectl get pod nginx-deployment-66b6c48dd5-dv6gl --subresource=scale
+Error from server (NotFound): the server could not find the requested resource
+```
+
+#### patch
+
+```shell=
+# For built-in types
+# update spec.replicas through scale subresource
+$ kubectl patch deployment nginx-deployment --subresource='scale' --type='merge' -p '{"spec":{"replicas":2}}'
+scale.autoscaling/nginx-deployment patched
+
+# spec.replicas is updated for the main resource
+$ k get deploy nginx-deployment
+NAME               READY   UP-TO-DATE   AVAILABLE   AGE
+nginx-deployment   2/2     2            2           4m
+
+# For CRDs
+$ kubectl patch crontabs cron --subresource='status' --type='merge' -p '{"status":{"replicas":2}}'
+crontab.stable.example.com/cron patched
+
+$ kubectl get crontab cron --subresource=scale
+NAME   DESIREDREPLICAS   AVAILABLEREPLICAS
+cron   3                 2
+```
+
+
+#### apply
+
+```shell=
+# cron-with-status.yaml has .status.replicas=3
+$ kubectl apply -f cron-with-status.yaml --subresource=status
+
+# .status.replicas is updated to 3
+$ kubectl get crontabs cron -o jsonpath='{.status.replicas}'
+3
+
+# cron-with-status.yaml updated to include .status.replicas=2
+# but is not applied if --subresource flag is not used
+$ kubectl apply -f cron-with-status.yaml
+
+# .status.replicas is not updated
+$ kubectl get crontabs cron -o jsonpath='{.status.replicas}'
+3
+```
+
+### Notes
+
+While this flag now allows updating `.status` for custom resources when a custom controller is not running against them, this is _not_ problematic because:
+
+- This feature is gated by the `--subresource` flag, and does not change the default behavior for kubectl.
+- Updating `.status` is also possible today by using `curl` directly.
```

Comment:
While this point may be true, making the 'status' field as first class entity in kubectl, especially for patch, apply and edit, can be problematic from usability perspective. It can lead to confusion on part of general users - what does it really mean to update a 'status' of a resource through kubectl? We will need to provide extensive explanation to the effect that if there is a controller that controls the status then the action will be a no-op, or a failure. Currently with the curl based option, any user who is wanting to update the status, is making conscious efforts and decisions to make such as update. In some sense we can assume that they are aware of what editing/patching a status field will mean (whether it will happen or will be a no-op, etc.)